src/compile/compile.ts

```diff
@@ -71,6 +71,16 @@ function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelPropert
   // Config with Vega-Lite only config removed.
   const vgConfig = model.config ? stripAndRedirectConfig(model.config) : undefined;
 
+  // Assemble data before projection
+  const data = [].concat(
```

Comment:
It should no longer cause side effect, so no longer need to move it.  (Or at least, remove the comment that it has to come before projection